### PR TITLE
Fixes for modern EC's, brightness scaling & threaded write support

### PR DESCRIPTION
**Pull request by SytheZN**
This PR includes 3 sets of changes; I suggest reviewing by commit to make understanding each easier.

[feat: refactor legacy and implement modern ec control](https://github.com/ShadowBlip/ayaneo-platform/pull/11/commits/88f82745d64e9b1a8d0cc120659c5957717b67e7)

* Renames the legacy write methods to `_legacy_`.
* Reworks the modern write methods to follow the same shape as legacy.
* Updates the commands sent to modern devices to align with device expectations.

[feat: add brightness scaling to match hardware capabilities](https://github.com/ShadowBlip/ayaneo-platform/pull/11/commits/adb9ba57eae237152c83fbdec3d37a487af7cc1f)

* Implements device specific scaling to match the device limits.

[feat: add threaded write support](https://github.com/ShadowBlip/ayaneo-platform/pull/11/commits/958ab5a27e339acd2ce7b1281b06b8f0a2e6f3ff)

* Moves the EC writes to a thread.
* When writes occur faster than the EC can handle, updates as fast as possible.

*Edit: missed a commit from main; rebased to fix

### CODE REVIEW

**Comment by pastaq** [comment]
This is pretty hard to follow. In the future I would prefer to keep each feature self contained into it's own PR. I understand in this case there was a lot of renamed functions which would have made that difficult. 

I'll review this in detail tonight.

**Comment by SytheZN** [comment]
Yeah, definitely understand that. I'm happy to split it up if it's easier - I had them in separate branches originally, but since the changes were built on top of the previous ones I figured it'd be a bit less testing.

File: ayaneo-platform.c [hunk 1]
```diff
@@ -266,139 +280,71 @@ static int write_ec_ram(u8 index, u8 val)
...
 {
-        u8 quadrant;
-
-        // Zone 1 (Left Stick)
-        ayaneo_led_mc_open();
-        for(quadrant = 0; quadrant < 12; quadrant++) {
-                write_ec_ram(0xB3 + quadrant, color[quadrant % 3]); // Quadrant 1
-        }
-        ayaneo_led_mc_write();
+        ec_write_ram(0xb2, 0xb1);
```
Comment (pastaq) [inline ayaneo-platform.c]: These two functions could be combined if we establish a const for 0xb1 and 0xc1

File: ayaneo-platform.c [hunk 2]
```diff
@@ -422,73 +368,140 @@ static void ayaneo_led_mc_set(u8 group, u8 pos, u8 brightness)
...
 
         for (zone = 0; zone < 4; zone++) {
-                ayaneo_led_mc_set(group, zones[zone], color[0]);
-                ayaneo_led_mc_set(group, zones[zone] + 1, color[1]);
-                ayaneo_led_mc_set(group, zones[zone] + 2, color[2]);
+                ayaneo_led_mc_legacy_set(group, zones[zone], color[0]);
+                ayaneo_led_mc_legacy_set(group, zones[zone] + 1, color[1]);
+                ayaneo_led_mc_legacy_set(group, zones[zone] + 2, color[2]);
         }
+        ayaneo_led_mc_legacy_set(0x03, 0x00, 0x00);
```
Comment (pastaq) [inline ayaneo-platform.c]: What does this additional set call do?

File: ayaneo-platform.c [hunk 3]
```diff
@@ -422,73 +368,140 @@ static void ayaneo_led_mc_set(u8 group, u8 pos, u8 brightness)
...
+}
+
```
Comment (pastaq) [inline ayaneo-platform.c:453]: Could you call this in the loop for ayaneo_led_mc_legacy_intensity?

File: ayaneo-platform.c [hunk 4]
```diff
@@ -422,73 +368,140 @@ static void ayaneo_led_mc_set(u8 group, u8 pos, u8 brightness)
...
+        ayaneo_led_mc_legacy_set(0x03, 0x00, 0x00);
+}
+
+static void ayaneo_led_mc_legacy_off(void)
+{
+        ayaneo_led_mc_legacy_set(0x01, 0x02, 0x31);
+        ayaneo_led_mc_legacy_set(0x02, 0x02, 0x31);
+
+        // note: omit for aya flip when implemented
```
Comment (pastaq) [inline ayaneo-platform.c]: Why?

**Review by pastaq — COMMENTED** [review]
Tested on the Air Plus AMD and 2s. Everything worked and the return time was about 3x faster on average which is really nice. Color was consistent right/left on all devices and there were no regressions to suspend.

My only request relates to the driver having got to a point where there are quite a few similarly named functions and it isn't always clear what exactly is going on anymore. If you could add some comments to the functions about what they do and why that would help the readability a bunch.

File: ayaneo-platform.c (hunk 2 above)
Comment (SytheZN) [reply ayaneo-platform.c]: I believe it signals the microcontroller to apply the change, kind of like a commit or CTS.

0x03 refers to both groups (0x01 & 0x02), left and right.

Without it I noticed the on/off commands occasionally failed to apply prior to suspend.

File: ayaneo-platform.c (hunk 3 above)
Comment (SytheZN) [reply ayaneo-platform.c:453]: Yeah, I've refactored that now, and added defines for a number of the commands as well to make their usage clearer.

File: ayaneo-platform.c (hunk 4 above)
Comment (SytheZN) [reply ayaneo-platform.c]: I believe the controller on the flip has a bug or quirk which causes unexpected behavior when it receives the commit call.
Updated comments to clarify.

File: ayaneo-platform.c (hunk 1 above)
Comment (SytheZN) [reply ayaneo-platform.c]: Yeah, I think on/off/reset could all be a single function, but I was trying to keep the intent behind each clear since they're used in various places.
I've introduced constants for the addresses and values at least so the writes make sense.
